snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -208,6 +212,10 @@ public List<SnowflakeIdentifier> listDatabases() {
                   queryHarness.query(
                       conn, "SHOW DATABASES IN ACCOUNT", DATABASE_RESULT_SET_HANDLER));
     } catch (SQLException e) {
+      // 2003, 2043, 2001 are possible Snowflake error codes for Object does not exist
+      if (e.getErrorCode() == 2003 || e.getErrorCode() == 2043 || e.getErrorCode() == 2001) {
+        throw new NoSuchNamespaceException("%s", e.getMessage());
```

Comment:
Wherever we throw a new exception, we should probably continue to provide the original cause as the `Throwable cause`. If a caller handles the exception cleanly it'll suppress the details of the re-thrown exception anyways, so for the cases that don't handle it cleanly it's better to preserve the full causal chain.

Here and below.